pkg/networkserver/redis/application_uplink_queue.go

```diff
@@ -82,119 +89,62 @@ var (
 // Subscribe ranges over q.upStream(unique.ID(ctx, appID)) using f until ctx is done.
 // Subscribe assumes that there's at most 1 active consumer in q.Group per stream at all times.
 func (q *ApplicationUplinkQueue) Subscribe(ctx context.Context, appID ttnpb.ApplicationIdentifiers, f func(context.Context, *ttnpb.ApplicationUp) error) error {
-	streamULID, err := ulid.New(ulid.Now(), rand.Reader)
-	if err != nil {
-		return err
-	}
 	uid := unique.ID(ctx, appID)
 	upStream := q.uidUplinkKey(uid)
-	closeStream := q.uidCloseKey(uid, streamULID.String())
 
-	_, err = q.Redis.XGroupCreateMkStream(closeStream, q.Group, "0").Result()
-	if err != nil {
+	_, err := q.Redis.XGroupCreateMkStream(upStream, q.Group, "0").Result()
+	if err != nil && !ttnredis.IsConsumerGroupExistsErr(err) {
 		return ttnredis.ConvertError(err)
 	}
 
-	dl, hasDL := ctx.Deadline()
-	if hasDL {
-		_, err = q.Redis.ExpireAt(closeStream, dl).Result()
-		if err != nil {
-			return ttnredis.ConvertError(err)
-		}
-	}
-
-	doneCh := make(chan struct{})
-	defer func() {
-		close(doneCh)
-	}()
-	go func() {
-		logger := log.FromContext(ctx).WithField("key", closeStream)
-		select {
-		case <-ctx.Done():
-			_, err := q.Redis.XAdd(&redis.XAddArgs{
-				Stream: closeStream,
-				Values: map[string]interface{}{"": ""},
-			}).Result()
-			if err != nil {
-				logger.WithError(err).Error("Failed to add message to Redis stream")
-				return
-			}
-			<-doneCh
-
-		case <-doneCh:
-		}
-
-		_, err := q.Redis.Del(closeStream).Result()
-		if err != nil {
-			logger.WithError(err).Error("Failed to delete Redis key")
-		}
-	}()
-
-	_, err = q.Redis.XGroupCreateMkStream(upStream, q.Group, "0").Result()
-	if err != nil && !ttnredis.IsConsumerGroupExistsErr(err) {
-		return ttnredis.ConvertError(err)
+	upCh := make(chan *ttnpb.ApplicationUp, 1)
+	_, ok := q.subscriptions.LoadOrStore(uid, upCh)
+	if ok {
+		panic(fmt.Sprintf("duplicate subscription for application %s", uid))
```

Comment:
This should error, not panic. `ApplicationUplinkQueue.Subscribe` is called by the implementation of the `LinkApplication` RPC, which will be called by an (external) Application Server.

Situation: When an Application Server restarts because of a crash, the old application link may not have been canceled yet (the TCP connection may not have been closed cleanly) when the new instance calls LinkApplication, that causes a `panic` 💥

Reply:
The previous link is always closed before a new one is opened - https://github.com/TheThingsNetwork/lorawan-stack/blob/d6724d1cfb463b4096207e829cbbeb2e6daafe33/pkg/networkserver/grpc_asns.go#L85-L91
This is only reachable when there's a bug in NS implementation